### DISCUSSION TITLE
APQ support with ApolloServer

### DISCUSSION THREAD

**Discussion by intellix** [enisdenjo/graphql-ws#445] [Q&A] [answered]
So I've randomly stumbled upon the example where you can use persistedQueries and it gave me the thirst for adding it to speed up socket messages. The example on the homepage is extremely barebones and I guess it just gives an idea that it's possible using graphql-ws, but I was wondering if anyone else had achieved it with a full ApolloClient/Server stack at all and could give some pointers for how to achieve it with Automatic Persisted Queries?

So with APQ the idea is that you mostly assume the persisted query is already known in the server and if it responds with: "PersistedQueryNotFound" then you remake the same query and provide the query, which is cached and then everyone follows the happy path. Without it you'd need some sort of build process for adding all possible persisted queries from the frontend and storing them in the backend which would be a ballache.

Websocket client in frontend:  

<img width="1026" alt="Screenshot 2023-02-09 at 14 37 25" src="https://user-images.githubusercontent.com/1162531/217844746-7cf8a5d4-b7e9-4336-9904-bc3a8b520cf2.png">

So in the client, I've basically done this which attempts to make the request with only the persistedQuery (at which point the backend is meant to respond that it doesn't exist):

```ts
import { createPersistedQueryLink } from 'apollo-angular/persisted-queries';
createPersistedQueryLink({ sha256 }).concat(websocketLink)
```

Which looks to work:  
<img width="1136" alt="Screenshot 2023-02-09 at 14 38 17" src="https://user-images.githubusercontent.com/1162531/217843303-1aa1befa-d8f6-4ee7-b255-15cff4f7a00a.png">

One part I'm not sure about, is if the retrying is automatically handled within the socket, but for now let's look at the server:

Unfortunately the PersistedQueryError isn't exposed so I just copied it from ApolloServer:  
```ts
export class PersistedQueryNotFoundError extends GraphQLErrorWithCode {
  constructor() {
    super('PersistedQueryNotFound', ApolloServerErrorCode.PERSISTED_QUERY_NOT_FOUND, {
      extensions: {
        http: {
          headers: new HeaderMap([['cache-control', 'private, no-cache, must-revalidate']]),
          status: 200,
        },
      },
    });
  }
}
```

And within graphql-ws useServer onSubscribe hook I'm searching for the query from the same cache where ApolloServer puts it and am throwing the required error to alert the frontend that it doesn't exist and you need to re-make the same query:
```ts
async onSubscribe({ extra }, msg) {
  let query: string;
  const pqh = msg.payload.extensions?.persistedQuery as { sha256Hash: string; version: string };

  if (pqh) {
    query = await cache.get(`apq:${pqh.sha256Hash}`);
    if (!query) {
      throw new PersistedQueryNotFoundError();
    }
  }
  
  const args = {
    document: parse(query ?? msg.payload.query),
    ...
```

Now my current problem is that when the socket receives an error, it just dies and I can't even see it in the response:  

<img width="679" alt="Screenshot 2023-02-09 at 14 41 51" src="https://user-images.githubusercontent.com/1162531/217844333-4804b892-5e51-4e19-8a3d-69888fe1eea3.png">

The error is basically a CloseEvent:  
<img width="923" alt="Screenshot 2023-02-09 at 14 44 38" src="https://user-images.githubusercontent.com/1162531/217845120-86ca7ad0-352d-44ad-98f4-2d81bc40da6e.png">

And the error on the server as well suggests I shouldn't be throwing inside onSubscribe():

<img width="405" alt="Screenshot 2023-02-09 at 14 59 51" src="https://user-images.githubusercontent.com/1162531/217849023-f2fec083-3911-4397-8bbe-17fb8f2cff02.png">

I'm guessing that for APQ I wouldn't be able to use Errors at all with sockets and I'd have to return the error as a standard response?

**Comment by enisdenjo** [comment, marked as answer]
Hey there, I am not an Apollo expert but I can help out with the graphql-ws part.

The reason why the whole connection closes is because graphql-ws server treats all errors thrown from hooks as fatal and unexpected. If you want to stop/error-out only one specific subscription, both client and server side, you should return an array of `GraphQLError`s from `onSubscribe` instead.

```diff
async onSubscribe({ extra }, msg) {
  let query: string;
  const pqh = msg.payload.extensions?.persistedQuery as { sha256Hash: string; version: string };

  if (pqh) {
    query = await cache.get(`apq:${pqh.sha256Hash}`);
    if (!query) {
-      throw new PersistedQueryNotFoundError();
+      return [new GraphQLError('Persisted query not found')]
    }
  }
  
  const args = {
    document: parse(query ?? msg.payload.query),
    ...
```

Since the error gets serialised when sent to the client, it doesn't really matter what type of it is.
